source/validate_adjacency.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2017 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Validates correctness of the intra-block preconditions of SPIR-V instructions.
+
+// TODO: Rename this file to validate_adjacency.cpp.
+
+#include "validate.h"
+
+#include <string>
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+spv_result_t ValidateAdjacency(ValidationState_t& _) {
+  auto const instructions = _.ordered_instructions();
+  for (auto i = instructions.cbegin(); i != instructions.cend(); ++i) {
+    switch (i->opcode()) {
+      case SpvOpPhi:
+        switch (prev(i)->opcode()) {
+          case SpvOpLabel:
+          case SpvOpPhi:
+          case SpvOpLine:
+            break;
+          default:
+            return _.diag(SPV_ERROR_INVALID_DATA)
+                   << "OpPhi must appear before all non-OpPhi instructions "
+                   << "(except for OpLine, which can be mixed with OpPhi).";
+        }
+        break;
+      case SpvOpLoopMerge:
+        switch (next(i)->opcode()) {
+          case SpvOpBranch:
+          case SpvOpBranchConditional:
+            break;
+          default:
+            return _.diag(SPV_ERROR_INVALID_DATA)
+                   << "OpLoopMerge must immediately precede either an OpBranch "
+                   << "or OpBranchConditional instruction. OpLoopMerge must be "
+                   << "the second-to-last instruction in its block.";
+        }
+        break;
+      case SpvOpSelectionMerge:
+        switch (next(i)->opcode()) {
```

Comment:
I'd like a case where OpSelectionMerge is the last instruction in the module.

Reply:
Done.